tests/integration/vxlan_test.py

```diff
@@ -0,0 +1,118 @@
+# the Free Software Foundation, either version 2.1 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+#
+
+import time
+
+import pytest
+
+import libnmstate
+from libnmstate.error import NmstateVerificationError
+from libnmstate.schema import Interface
+from libnmstate.schema import InterfaceState
+from libnmstate.schema import InterfaceType
+
+from .testlib import assertlib
+from .testlib.vxlan import apply_state
+from .testlib.vxlan import vxlan_down
+from .testlib.vxlan import vxlan_interfaces
+from .testlib.vxlan import vxlan_up
+from .testlib.vxlan import vxlans_absent
+from .testlib.vxlan import vxlans_up
+
+VXLAN1_ID = 101
+VXLAN2_ID = 102
+
+
+def test_add_and_remove_vxlan(eth1_up):
+    ifname = eth1_up[Interface.KEY][0][Interface.NAME]
+    with vxlan_interfaces(
+        [(VXLAN1_ID, ifname, '192.168.100.1')]
+    ) as desired_state:
+        assertlib.assert_state(desired_state)
+
+    vxlan1_ifname = desired_state[Interface.KEY][0][Interface.NAME]
+    assertlib.assert_absent(vxlan1_ifname)
+
+
+def test_add_and_remove_two_vxlans_on_same_iface(eth1_up):
+    ifname = eth1_up[Interface.KEY][0][Interface.NAME]
+    with vxlan_interfaces(
+        [
+            (VXLAN1_ID, ifname, '192.168.100.1'),
+            (VXLAN2_ID, ifname, '192.168.100.2'),
+        ]
+    ) as desired_state:
+        assertlib.assert_state(desired_state)
+
+    vxlan_interfaces_name = (
+        i[Interface.NAME] for i in desired_state[Interface.KEY]
+    )
+    assertlib.assert_absent(vxlan_interfaces_name)
+
+
+def test_rollback_for_vxlans(eth1_up):
+    ifname = eth1_up[Interface.KEY][0][Interface.NAME]
+    current_state = libnmstate.show()
+    desired_state = {
+        Interface.KEY: [
+            vxlan_up(VXLAN1_ID, ifname, '192.168.100.1'),
+            vxlan_up(VXLAN2_ID, ifname, '192.168.100.2'),
+        ]
+    }
+
+    desired_state[Interface.KEY][1]['invalid_key'] = 'foo'
+    with pytest.raises(NmstateVerificationError):
+        libnmstate.apply(desired_state)
+
+    time.sleep(5)  # Give some time for NetworkManager to rollback
+    current_state_after_apply = libnmstate.show()
+    assert current_state == current_state_after_apply
+
+
+def test_set_vxlan_iface_down(eth1_up):
+    ifname = eth1_up[Interface.KEY][0][Interface.NAME]
+    with vxlan_interfaces(
+        [(VXLAN1_ID, ifname, '192.168.100.1')]
+    ) as desired_state:
+        libnmstate.apply({Interface.KEY: [vxlan_down(VXLAN1_ID, ifname)]})
+
+    vxlan_ifname = desired_state[Interface.KEY][0][Interface.NAME]
+    assertlib.assert_absent(vxlan_ifname)
```

Comment:
Don't you need to check this inside the contextmanager block? At this point, it already clean-up.

Reply:
Quite the oppsite, it checks that interface is not there after it's out of the scope of context manager since it will call absent out of context.